### PR TITLE
Add arrow module in Py API

### PR DESCRIPTION
**Pull request by jmao-denver**
Enable conversion between DH tables and Arrow tables on the server side

### CODE REVIEW

**Comment by jmao-denver** [comment]
The 'missing pyarrow' fixes should be removed once the base images are updated with 'pyarrow' preinstalled as with numpy/pandas etc.

**Review by devinrsmith — DISMISSED** [review]
This can't merge as is - the `--no-index` is a "safety" check that ensures everything flows through the base images.

If this is necessary, let's plumb https://github.com/deephaven/deephaven-base-images.

pyarrow is not a small dependency - it's 20-35MB depending on architecture, and it has a dependency on numpy.

Note, pyarrow is an architecture specific deliverable - but it looks like it has good coverage w/ amd64 and arm64 across Linux, Mac, and Windows with python 3.7-3.11.

File: py/server/tests/test_arrow.py
Comment (chipkent) [inline py/server/tests/test_arrow.py:83]: These tests are all deephaven-centric.  We need another test that tests all supported and unsupported arrow types (e.g. utf8).

**Comment by jmao-denver** [comment]
@nbauernfeind @lbooker42 has looked that the changes in Java and thought they are reasonable. Could you also review these changes?

File: py/server/tests/test_arrow.py
Comment (nbauernfeind) [reply py/server/tests/test_arrow.py:22]: Boolean support is officially up for review now. This likely works on top of that change.

After a merge + rebase of boolean support we should bolster the tests below to include boolean columns, too.

File: py/server/tests/test_arrow.py
Comment (nbauernfeind) [inline py/server/tests/test_arrow.py:46]: This test feels like a land mine. If `to_table` creates an empty table then this test will pass, but would clearly be incorrect.

Could we compare the converted table to the version that our ParquetTools reader would generate instead?

File: py/server/tests/test_arrow.py
Comment (nbauernfeind) [inline py/server/tests/test_arrow.py:84]: Hmm, is there a reason we want to expect this? Is the issue that the file has boolean columns or some other type that is mismatching somehow?

File: py/server/deephaven/arrow.py
Comment (nbauernfeind) [reply py/server/deephaven/arrow.py:104]: > but it ends up being excessively complex and inefficient

Unfortunately we do not use Arrow Buffers internal to DHC. The alternative implementation that avoids copying data twice requires building column source wrappers that are backed by Arrow's native buffers. If we want to support converting a DH native table (e.g. converting a ticking table to PyArrow's format) then we must do the copying. Note that our column sources support all sorts of fancy features like redirecting to yield a sorted table.

Our internal column source organization is specifically what makes pulling py arrow data into DH an expensive operation (as well as vice versa).

**Comment by jcferretti** [comment]
From the point of view of the question "is there overlap/work with the C++ wrapping / python client", I think the answer is not directly, and perhaps a way to understand that question would be to ask "should we worry about API consistency from the point of view of exposing arrow calls in python-server, python-client and C++-client?
I think that is a fair concern and we can look into that but my guess is right now is not the best time to do that, we have an open question on the C++ wrapping side about exactly how arrow is going to be exposed for python when we integrate both C++ libraries that depend on arrow code and python code that depends on pyarrow.  Without having clarity on the implementation level potential restrictions there, we can't (shouldn't?) say much now I think.
So I think is better to move this PR forward as is and perhaps open a separate ticket to check arrow API consistency between py-server, py-client and C++-client later.

File: py/server/deephaven/arrow.py
Comment (jmao-denver) [reply py/server/deephaven/arrow.py:71]: So far our convention is to not use Optional. If we decide to adopt the use of Optional then it should be a wholesale change across the package.

File: py/server/deephaven/arrow.py
Comment (jmao-denver) [reply py/server/deephaven/arrow.py:85]: This will allocate a bit memory for the new table's metadata but not for the data, similar to what we do in DH.

File: py/server/tests/test_arrow.py
Comment (jmao-denver) [reply py/server/tests/test_arrow.py:46]: Good point, added another assert on table size to tighten the test a bit.

File: py/server/tests/test_arrow.py
Comment (jmao-denver) [reply py/server/tests/test_arrow.py:84]: this is just a placeholder for a potential bug we ran into, changed the name to make it more obvious.

File: py/server/tests/test_arrow.py
Comment (jmao-denver) [reply py/server/tests/test_arrow.py:46]: Also added a test to specifically check against the table read in through DH's parquet tool

**Comment by jmao-denver** [comment]
#3223 is created to track the lack of support of Arrow date32() type as captured in one of the unit tests.

File: py/server/deephaven/arrow.py
Comment (jmao-denver) [reply py/server/deephaven/arrow.py:68]: 1. yea, we could make it part of dtypes, but I prefer to wait until we support more arrow data types and/or have some ways to use arrow data directly in DHC Python API
2. the ''  is really there as a place holder to keep the list complete and indicates that we should support the type in the future

**Review by nbauernfeind — APPROVED** [review]
LGTM if looks good to @chipkent

File: py/server/setup.py
Comment (chipkent) [inline py/server/setup.py:60]: We should have a tracking ticket on our side to resolve this.

File: py/server/deephaven/arrow.py
Comment (chipkent) [inline py/server/deephaven/arrow.py:44]: Should we map to our Periods or to ns?

File: py/server/deephaven/arrow.py
Comment (chipkent) [inline py/server/deephaven/arrow.py:33]: Any reason these can't be mapped to DateTime?  Are we only supporting types that have full round-trip?  Are we not supporting type mapping?  etc.

File: py/server/deephaven/arrow.py
Comment (chipkent) [inline py/server/deephaven/arrow.py:50]: Can this not just be a `java.lang.String` on the DH side?

File: py/server/deephaven/arrow.py
Comment (chipkent) [inline py/server/deephaven/arrow.py:75]: 1) Together, these two are inconsistent.  We do support timestamps if they have timezones, but that is not reflected in `SUPPORTED_ARROW_TYPES`.

2) Are we not supporting cases without timezones to force users to specify the timezone or for some other reason?

3) I think that no specified timezone is traditionally interpreted as the machine local timezone.  https://arrow.apache.org/docs/python/timestamps.html

File: py/server/deephaven/arrow.py
Comment (chipkent) [reply py/server/deephaven/arrow.py:109]: It seems like Arrow->DH necessarily creates a static DH table.  In that case, we should have column source support for Arrow buffers.  That is certainly a follow-on project, but the inefficiency here isn't great.  At a minimum, we should have a ticket for this effort.  The copy will certainly make some users unhappy at some point.

File: py/server/deephaven/arrow.py
Comment (chipkent) [inline py/server/deephaven/arrow.py:144]: Does this handle UGP locking for ticking tables, or does the method need a locking annotation?  If locking is happening, there should be a comment, since it isn't transparent.

File: py/server/tests/test_arrow.py
Comment (chipkent) [inline py/server/tests/test_arrow.py:88]: This looks like a problem.  We have a number of supported data types that have disabled unit tests and apparently do not work.

File: py/server/deephaven/arrow.py
Comment (chipkent) [inline py/server/deephaven/arrow.py:38]: As noted later, it does look like we support timestamps if a timezone is present.

**Review by chipkent — CHANGES_REQUESTED** [review]

File: py/server/deephaven/arrow.py
Comment (jmao-denver) [reply py/server/deephaven/arrow.py:44]: @nbauernfeind Your thoughts?

File: py/server/deephaven/arrow.py
Comment (jmao-denver) [reply py/server/deephaven/arrow.py:33]: @nbauernfeind?

File: py/server/deephaven/arrow.py
Comment (jmao-denver) [reply py/server/deephaven/arrow.py:50]: @nbauernfeind ?

File: py/server/deephaven/arrow.py
Comment (jmao-denver) [reply py/server/deephaven/arrow.py:144]: 1. #3273 is created for the possible follow-on project.
2. The engine code handles the locking for ticking table efficiently and automatically,  nothing needs to be done in Python wrapper code.

File: py/server/deephaven/arrow.py
Comment (nbauernfeind) [reply py/server/deephaven/arrow.py:44]: We can, but this requires a separate set of java changes and should not be bundled here.

File: py/server/deephaven/arrow.py
Comment (nbauernfeind) [reply py/server/deephaven/arrow.py:33]: We can, but this requires a separate set of java changes and should not be bundled here.

Currently, we only support types that can be round-tripped and have a unique on-to mapping. To support multiple source types mapping to the same destination type requires java-parsing-side effort.

File: py/server/deephaven/arrow.py
Comment (nbauernfeind) [reply py/server/deephaven/arrow.py:50]: We probably can, but this requires a separate set of java changes and should not be bundled here.

File: py/server/deephaven/arrow.py
Comment (nbauernfeind) [reply py/server/deephaven/arrow.py:109]: There is DHE work (DH-12088) that supports columns that wrap arrow buffers. I'm tasked to port that, but from a 10k ft view, it's still quite a bit of work.

File: py/server/tests/test_arrow.py
Comment (nbauernfeind) [reply py/server/tests/test_arrow.py:88]: I appreciate having tests that I can enable once the java side supports new (or more-correct) conversions. Let's leave them in.

File: py/server/deephaven/arrow.py
Comment (nbauernfeind) [reply py/server/deephaven/arrow.py:38]: It's possible that we only support the `ns` resolution version of this timestamp.

File: py/server/setup.py
Comment (jmao-denver) [reply py/server/setup.py:60]: I have filed a ticket with Turbodbc and have helped find a solution to the problem.  I will be notified once the fix is merged and tagged for release.

File: py/server/deephaven/arrow.py
Comment (jmao-denver) [reply py/server/deephaven/arrow.py:75]: 1. as one of the skipped test cases shows, the Arrow->DH conversion of timestamp isn't done correctly. The problem is tracked by #3223. So the SUPPORTED_ARROW_TYPES is correct even though the issue with 'timestamp' isn't strictly 'not supported'.

File: py/server/deephaven/arrow.py
Comment (chipkent) [reply py/server/deephaven/arrow.py:44]: This looks like it is covered in https://github.com/deephaven/deephaven-core/issues/3223

File: py/server/deephaven/arrow.py
Comment (chipkent) [reply py/server/deephaven/arrow.py:33]: This looks like it is covered in https://github.com/deephaven/deephaven-core/issues/3223

File: py/server/deephaven/arrow.py
Comment (chipkent) [reply py/server/deephaven/arrow.py:75]: I have concrerns that `timestamps` return values, but the values are wrong.  Users are not very tolerant of corrupt data.

File: py/server/deephaven/arrow.py
Comment (chipkent) [reply py/server/deephaven/arrow.py:38]: This looks like it is covered in https://github.com/deephaven/deephaven-core/issues/3223

**Review by chipkent — APPROVED** [review]

**Review by nbauernfeind — APPROVED** [review]

**Review by chipkent — APPROVED** [review]

**Review by devinrsmith — APPROVED** [review]
Docker / setup.py additions look correct to me; haven't reviewed the arrow-specific parts.

**Pull request merged**
